packages/flutter_tools/lib/src/version.dart

```diff
@@ -752,81 +752,117 @@ class GitTagVersion {
         _runGit('git fetch $_flutterGit --tags', processUtils, workingDirectory);
       }
     }
-    // `--match` glob must match old version tag `v1.2.3` and new `1.2.3-dev.4.5`
-    return parse(_runGit('git describe --match *.*.* --first-parent --long --tags', processUtils, workingDirectory));
+    final List<String> tags = _runGit(
+      'git tag --contains HEAD', processUtils, workingDirectory).split('\n');
+
+    // Check first for a stable tag
+    final RegExp stableTagPattern = RegExp(r'^\d+\.\d+\.\d+$');
+    for (final String tag in tags) {
+      final String trimmedTag = tag.trim();
+      if (stableTagPattern.hasMatch(trimmedTag)) {
+        return parse(trimmedTag);
+      }
+    }
+    // Next check for a dev tag
+    final RegExp devTagPattern = RegExp(r'^\d+\.\d+\.\d+-\d+\.\d+\.pre$');
+    for (final String tag in tags) {
+      final String trimmedTag = tag.trim();
+      if (devTagPattern.hasMatch(trimmedTag)) {
+        return parse(trimmedTag);
+      }
+    }
+
+    // If we're not currently on a tag, use git describe to find the most
+    // recent tag and number of commits past.
+    return parse(
+      _runGit(
+        'git describe --match *.*.*-*.*.pre --first-parent --long --tags',
+        processUtils,
+        workingDirectory,
+      )
+    );
   }
 
-  // TODO(fujino): Deprecate this https://github.com/flutter/flutter/issues/53850
-  /// Check for the release tag format of the form x.y.z-dev.m.n
-  static GitTagVersion parseLegacyVersion(String version) {
+  /// Check for the release tag format of the forms x.y.z or x.y.z-m.n.pre
+  static GitTagVersion parseTaggedVersion(String version) {
```

Comment:
`parseTaggedVersion` and `parseGitVersion` look almost the same.  Can they be combined with a regex that handles either case?  `?>` is a non-capturing group.  
I didn't run the unit tests, so that might not be quite right.

^(\d+)\.(\d+)\.(\d+)(-\d+\.\d+\.pre)?(?>-(\d+)-g([a-f0-9]+))?$

Reply:
We could combine the regexes, but i originally split these into two different methods as a code review suggestion because the the regex was getting difficult to read (which I agree). In particular, I feel like more explicit the regex, the less likely it is to be wrong.

At the same time, it feels bad having two methods that are almost the same. So not sure what to do.

Reply:
The methods seem sufficiently complex at this point to deduplicate, I'd rather trade an evil regex for fewer potential copy-paste issues like:
```
      m = devGroups[0] == null ? null : int.tryParse(devGroups[0]);
      n = devGroups[1] == null ? null : int.tryParse(devGroups[1]);
```

Reply:
Ok, I collapsed it into one. PTAL